### PR TITLE
Write-locked graph storage implementation for faster bulk loaders

### PR DESCRIPTION
**Pull request by ljeub-pometry**
### What changes were proposed in this pull request?

Adds write-locked views for the underlying graph storage which should enable faster bulk loaders

### Issues

_If this resolves any issues, please link to them here, the format is a KEYWORD followed by @<ISSUE NUMBER>__
KEYWORDS available are `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves`, `resolved`.
Please delete this text before creating your PR

### CODE REVIEW

File: raphtory/src/core/storage/mod.rs [hunk 1]
```diff
@@ -298,6 +296,81 @@ where
...
+    pub fn set(&mut self, vid: VID, gid: GidRef) {
+        if let Some(offset) = self.resolve(vid) {
+            if offset >= self.shard.len() {
+                self.shard.resize_with(offset + 1, NodeStore::default);
+            }
+            self.shard[offset] = NodeStore::resolved(gid.to_owned(), vid);
+        }
+    }
+
+    fn reserve(&mut self, new_global_len: usize) {
```
Comment (fabianmurariu) [inline raphtory/src/core/storage/mod.rs:341]: does not seem to be used?

File: raphtory/src/core/storage/mod.rs [hunk 2]
```diff
@@ -298,6 +296,81 @@ where
...
+            .into_par_iter()
+            .enumerate()
+            .map(move |(shard_id, shard)| NodeShardWriter {
+                shard,
+                shard_id,
+                num_shards,
+            })
+    }
+
+    pub fn reserve(&'a mut self, new_len: usize) {
```
Comment (fabianmurariu) [inline raphtory/src/core/storage/mod.rs:370]: doesn't seem to be used

File: raphtory/src/db/api/mutation/internal/internal_addition_ops.rs [hunk 3]
```diff
@@ -127,6 +145,11 @@ impl<G: DelegateAdditionOps> InternalAdditionOps for G {
         self.graph().resolve_node(n)
     }
 
+    #[inline]
```
Comment (fabianmurariu) [inline raphtory/src/db/api/mutation/internal/internal_addition_ops.rs:148]: this is not called by anything on any impl

File: raphtory/src/db/api/mutation/internal/internal_addition_ops.rs [hunk 4]
```diff
@@ -24,6 +29,9 @@ pub trait InternalAdditionOps {
     /// map external node id to internal id, allocating a new empty node if needed
     fn resolve_node<V: AsNodeRef>(&self, id: V) -> Result<MaybeNew<VID>, GraphError>;
 
+    /// map external node id to internal id, creating a new mapping if the node does not exist without allocating the node storage
+    fn resolve_node_no_init(&self, id: GidRef) -> Result<MaybeNew<VID>, GraphError>;
```
Comment (fabianmurariu) [inline raphtory/src/db/api/mutation/internal/internal_addition_ops.rs:33]: this is not called anywhere by any impl

File: raphtory/src/io/arrow/df_loaders.rs
Comment (fabianmurariu) [inline raphtory/src/io/arrow/df_loaders.rs:1]: I'd like a proptest where we generate edges and add them to a Graph via the usual way then load them via the loader.
Then we assert they match.

File: Cargo.toml [hunk 5]
```diff
@@ -119,7 +119,7 @@ wasm-bindgen-test = "0.3.41"
 memmap2 = { version = "0.9.4" }
 ahash = { version = "0.8.3", features = ["serde"] }
 strum = { version = "0.26.1", features = ["derive"] }
-bytemuck = { version = "1.15.0" }
+bytemuck = { version = "1.15.0", features = ["derive"] }
```
Comment (fabianmurariu) [inline Cargo.toml:122]: this is now at 0.18

File: raphtory/Cargo.toml [hunk 6]
```diff
@@ -73,6 +73,7 @@ pometry-storage = { workspace = true, optional = true }
 prost = { workspace = true, optional = true }
 prost-types = { workspace = true, optional = true }
 
+
 [target.'cfg(target_os = "macos")'.dependencies]
 snmalloc-rs = { workspace = true }
```
Comment (fabianmurariu) [inline raphtory/Cargo.toml:78]: are we going with this one? it's the least supported option (perhaps we should create an issue about this)

Comment (ljeub-pometry) [reply raphtory/Cargo.toml:78]: We should probably raise an issue (I had a go at reproducing it with a simple example but couldn't manage). This one is noticeably faster than jemalloc for this particular scenario so I left it for now.

File: raphtory/src/core/entities/nodes/node_store.rs [hunk 7]
```diff
@@ -31,19 +32,29 @@ pub struct NodeStore {
...
 
 impl NodeStore {
-    pub fn new(global_id: GID, t: TimeIndexEntry) -> Self {
-        let mut layers = Vec::with_capacity(1);
-        layers.push(Adj::Solo);
+    #[inline]
+    pub fn is_initialised(&self) -> bool {
+        self.vid != VID::default()
+    }
+    pub fn new(vid: VID) -> Self {
```
Comment (fabianmurariu) [inline raphtory/src/core/entities/nodes/node_store.rs:39]: this is stil not used

File: raphtory/Cargo.toml (hunk 6 above)
Comment (fabianmurariu) [reply raphtory/Cargo.toml:78]: Fine, long term I don't know what is the commitment of Microsoft to supporting this allocator

File: raphtory/src/core/entities/nodes/node_store.rs (hunk 7 above)
Comment (ljeub-pometry) [reply raphtory/src/core/entities/nodes/node_store.rs]: it is gone

**Review by fabianmurariu — APPROVED** [review]
LGTM

**Pull request merged**